internal/server/providers/azure.go

```diff
@@ -0,0 +1,143 @@
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"gorm.io/gorm"
+
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+)
+
+const (
+	graphGroupMemberEndpoint = "https://graph.microsoft.com/v1.0/me/memberOf"
+	graphGroupDataType       = "#microsoft.graph.group"
+)
+
+type graphObject struct {
+	Type        string `json:"@odata.type"`
+	DisplayName string `json:"displayName"`
+}
+
+type graphResponse struct {
+	Context string        `json:"@odata.context"`
+	Value   []graphObject `json:"value"`
+}
+
+type azure struct {
+	OIDC OIDC
+}
+
+func (a *azure) Validate() error {
+	return a.OIDC.Validate()
+}
+
+func (a *azure) ExchangeAuthCodeForProviderTokens(code string) (rawAccessToken, rawRefreshToken string, accessTokenExpiry time.Time, email string, err error) {
+	return a.OIDC.ExchangeAuthCodeForProviderTokens(code)
+}
+
+func (a *azure) RefreshAccessToken(providerUser *models.ProviderUser) (accessToken string, expiry *time.Time, err error) {
+	return a.OIDC.RefreshAccessToken(providerUser)
+}
+
+func (a *azure) GetUserInfo(providerUser *models.ProviderUser) (*InfoClaims, error) {
+	return a.OIDC.GetUserInfo(providerUser)
+}
+
+func (a *azure) SyncProviderUser(db *gorm.DB, user *models.Identity, provider *models.Provider) error {
+	providerUser, err := data.GetProviderUser(db, provider.ID, user.ID)
+	if err != nil {
+		return err
+	}
+
+	if err := checkRefreshAccessToken(db, providerUser, a); err != nil {
+		return fmt.Errorf("oidc sync failed to check users access token: %w", err)
+	}
+
+	// this checks if the user still exists
+	_, err = a.OIDC.GetUserInfo(providerUser)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return fmt.Errorf("%w: %s", internal.ErrBadGateway, err.Error())
+		}
+		return fmt.Errorf("could not get user info from provider: %w", err)
+	}
+
+	newGroups, err := checkMemberOfGraphGroups(string(providerUser.AccessToken))
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return fmt.Errorf("%w: %s", internal.ErrBadGateway, err.Error())
+		} else if errors.Is(err, ErrUnauthorized) {
+			newGroups = &[]string{} // set the groups empty to clear them
```

Comment:
If the Azure client isn't configured to allow getting groups from the graph API clear the user groups, log a warning, and continue.